### ISSUE TITLE
Bash-completion support

### ISSUE THREAD

**Issue by kaiserkyu** [tpm2-software/tpm2-tools#1049]
I'm working on bash completion for the tools, at this time i managed to obtain the options through 
the short summary that the tools print when they are execute without options, but, some tools wont print it (usage information)  because they got defined actions that take no options, is there a way to obtain this short summary for the missing tools? the idea of this feature was to get the options dynamically in order to be friendly to the tools that are not yet implemented, i will be grateful for any suggestion.

**Comment by williamcroberts** [comment]
That's a good approach. So this would apply to tools that don't take options outside of the common options? Perhaps we should dump usage output if you invoke the tool in a way that causes a non-zero exit status. However, if a tool can be invoked with no options/arguments, like tpm2_pcrlist, then it wouldn't help in that case. Perhaps find a common, unused option to force a tool to just print it's summary?

1. We know the options
2. We can infer if it takes arguments based on if the on_args handler is set, but we won't know much about how many arguments, or what they mean.

**Comment by martinezjavier** [comment]
> Perhaps find a common, unused option to force a tool to just print it's summary?

@williamcroberts that's exactly what I thought too.

It's a pity that at the time we didn't think about bash completion, for example we could had chosen to use `-h` to print the short text summary and `--help` to start the man page or fallback to the short summary if the man page isn't found.

**Comment by martinezjavier** [comment]
> like tpm2_pcrlist, then it wouldn't help in that

I think it would be a problem too for tools that default to read from the stdin (i.e: tpm2_hash), so is not only the ones that takes no options/arguments.

**Comment by martinezjavier** [comment]
@kaiserkyu thanks for working on this BTW! I think it would be very useful specially since we have a lot of tools and many, many options.

Could you please share your WIP so people can try / test it? That way I think that would be easier to understand the limitations and possible ways to move forward.

**Comment by kaiserkyu** [comment]
@williamcroberts the problem is with the tools that can be invoked with no options/arguments, by the time it works with the ones that print the short summary when invoking them with no options, i was thinking on @martinezjavier idea, use -h for the short summary but it depends of what you think because i wasn't sure if modifying the behavior of the options would be a good idea. There are thirty one tools that generates the options flawlessly and sixteen that have the issue when invoking them.

**Comment by martinezjavier** [comment]
@kaiserkyu what @williamcroberts suggested is to find a common unused option that could be used to force printing the short summary, I think that's the best approach.

The problem with using `-h` is that it's a non-backward compatible change. BTW, do you plan to add this change to the master branch / upcoming tpm2-tools 4.0 release or the stable release? Because the constraints are not the same, in the former we should maintain backward compatibility but not in the latter.

**Comment by kaiserkyu** [comment]
this is what i have right now, it is not too fancy but i will be adding more things in an incremental manner
```
_tpm2_tools()
{



    local cur prev words cword
    _init_completion || return
    local common_options=(-h --help -v --version -V --verbose -Q --quiet -Z --enable-errata)
    local aux1=$( ${COMP_WORDS[0]} )
    local aux2=$( echo "${aux1}" | tr "[]|" " " | awk '{if(NR>2)print}')
    IFS=' ' eval 'suggestions=($aux2)'
    suggestions=("${suggestions[@]}" "${common_options[@]}") #generate all the opts for the tool

    if [[  "${COMP_WORDS[0]}" == "tpm2_activatecredential" ]]; then # exameple
      case $prev in
          -f)
              _filedir
              return
              ;;
      esac
    fi

    if [[ "${COMP_WORDS[0]}" == "tpm2_hmac" ]]; then # example for some options's arguments
      case $prev in
        -g)
            suggestions=( $( compgen -W '0x4 sha1 0xB sha256 0xC sha384 0xD sha512 0x12 sm3_256' -- "$cur" ) )
            COMPREPLY=("${suggestions[@]}")
            return
      esac
    fi



    if [[ "$cur" == -* ]]; then #start completion

        COMPREPLY=( $( compgen -W '$( echo ${suggestions[@]//<value>} )' -- "$cur" ) )
        [[ $COMPREPLY == *= ]] && compopt -o nospace
        return
    fi

}
  #to obtain the installation path of the tools, it is necessary to know just one of them
  #and then assign the completion function to all the tools
  tools_for_completions=( $(find $( dirname $(which tpm2_import) ) -type f -printf '%f\n' | grep 'tpm2_') )
  for i in "${tools_for_completions[@]}"
  do
   complete -F _tpm2_tools $i
  done
```

**Comment by kaiserkyu** [comment]
@martinezjavier what kind of backward compatibility? it need to be constantly updated to maintain consistency?

**Comment by kaiserkyu** [comment]
i was thinking on add the option to complete algorithms and other arguments for some options like i did with tpm2_hmac and tpm2_activatecredential in the sample code above, but i will need to filter the tools because some options have different actions for each tool

**Comment by martinezjavier** [comment]
> @martinezjavier what kind of backward compatibility? it need to be constantly updated to maintain consistency?

For stable releases, we shouldn't change the options behavior or semantics.

**Comment by williamcroberts** [comment]
@kaiserkyu 
1. Anything on a stable branch, you can preety much always add an option, but you can't change antything
2. For master, we are currently 100% OK with hammering on options and getting them to be common, for at least groups of tools. Tools can be grouped on their command set in relation to the spec. Ie the spec lays out the different groups.
    1. Options are changing on master, so we may be causing you pain. Sorry. I really like the direction you took using the option output.

**Comment by martinezjavier** [comment]
@kaiserkyu I wanted to share my thoughts on this. First, you need to decide which branch to target this feature. I first thought about master but we are now near to have a 3.1 release that's the 3.X branch forward ported to tpm2-tools 2.0.0-rc. So I think it would be good to have this feature there and I believe it's possible without breaking backward compatibility (more on that below).

Second is how to force all the tools to print the short summary text so you can parse it. I think that the best approach would be to add an optional argument to the `-h, --help` option. This can be achieved by adding a `::` next to `h` in the optstring.

So we could for example have `--help summary` and `--help manpage`. If you don't pass any argument (`--help`), then the default would be `--help manpage`. That way you maintain backward compatibility with the previous versions of the tools.

You should keep the current behaviour of the tools that don't take options or expect to read something from the standard input to not print the short summary when executed without options, but this tools now can take `-h summary` to force printing the short summary text.

@williamcroberts and @joshuagl do you think the above is sensible?

**Comment by williamcroberts** [comment]
Closed in #1100, also dup of #589
